microcosm_eventsource/event_types.py

```diff
@@ -0,0 +1,154 @@
+from microcosm_eventsource.transitioning import nothing
+
+
+class EventTypeInfo(object):
+    """
+    Event type meta data.
+
+    """
+    def __init__(self, follows=None, accumulate=None, restarting=False, requires=()):
+        """
+        :param follows:    an instance of the event mini-grammar
+        :param accumulate:  whether the event type should accumulating state
+        :param restarting: whether the event type may restart a (new) version
+
+        """
+        self.follows = follows or nothing()
+        self.accumulate = accumulate or current()
+        self.restarting = restarting
+        self.requires = requires
+
+
+class EventType(Enum):
+    """
+    Based event type enum.
+
+    """
+    @property
+    def is_initial(self):
+        """
+        Can this event type be the initial event?
+
+        """
+        return not bool(self.value.follows)
+
+    @property
+    def is_accumulating(self):
+        """
+        Does this event type accumulate state?
+
+        """
+        return self.value.accumulating
+
+    @property
+    def is_restarting(self):
+        """
+        Does this event type restart a new version?
+
+        """
+        return self.value.restarting
+
+    @classmethod
+    def required_column_names(cls):
+        return {
+            column_name
+            for event_type in cls
+            for column_name in event_type.value.requires
+        }
+
+    @classmethod
+    def requires(cls, column_name):
+        return [
+            event_type
+            for event_type in cls
+            if column_name in event_type.value.requires
+        ]
+
+    def may_transition(self, state):
+        """
+        Can this event type transition from the given state?
+
+        :param state: a set of event types
+
+        """
+        if not state:
+            # only initial event types may come first
+            return self.is_initial
+
+        return self.value.follows(self.__class__, state)
+
+    def validate_transition(self, state):
+        """
+        Asssert that a transition is legal from the given state.
+
+        :param state: a set of event types
+
+        """
+        if self.may_transition(state):
+            return
+
+        if not state:
+            # event may not be initial
+            raise IllegalInitialStateError("Event type '{}' may not be initial".format(
+                self.name,
+            ))
+
+        # event may not follow previous
+        raise IllegalStateTransitionError("Event type '{}' may not follow [{}]".format(
+            self.name,
+            ", ".join(event_type.name for event_type in state),
+        ))
+
+    def accumulate_state(self, state):
+        """
+        Accumulate state.
+
+        """
+        return self.value.accumulate(state, self)
+
+    def next_version(self, version):
+        """
+        Compute next version.
+
+        """
+        if version is None:
+            return 1
+        if self.is_restarting:
```

Comment:
I think the versioning logic allows it but the transitioning logic might not.

 - The first version is always 1
 - If the first event were of a restarting type, it would still start at version 1
 - If the restarting type came later, you'd get version N+1

I don't think it's easy to express "initial or [some other condition]" yet as a state transition.